### ISSUE TITLE
how to delete a cluster safely?

### ISSUE THREAD

**Issue by tolgaio** [Altinity/clickhouse-operator#1162]
Hello,

I have a cluster set up with a certain layout configuration, utilizing a ReplicatedMergeTree in the cluster. I now wish to modify the layout to accommodate a different layout (adding/removing shards and replicas), OR completely deleting this cluster definition and creating a new one with a new name. 

However, when I do this change in the `ClickHouseInstallation` configuration, the new cluster is still trying to connect to the old cluster with the old name, and therefore, I'm seeing a lot of `DNSResolver: Cannot resolve host` errors in the log. Say for example, I have the following cluster definition in the `ClickHouseInstallation`:

```
...
clusters:
  - name: cluster1
    shards: 1
    replicas: 3
....
```

now, I basically want to change the layout and/or the name of this cluster. let's say,  I want to bring down the replica to `2` with this:

```
...
clusters:
  - name: cluster1
    shards: 1
    replicas: 2
...
```

What happens is that the replica-3 gets deleted, however, now replica 1 and 2 are all trying to connect to the replica3 somehow. there are still traces of replica 3, ad the error log `DNSResolver: Cannot resolve host ...` suggests that it's trying to connect to replica3 which doesn't exist anymore.

Same goes the same for changing the name of the cluster. Let's say I'm changing the name while keeping the layout the same:

```
...
clusters:
  - name: cluster2
    shards: 1
    replicas: 3
...
```

the old pods get deleted, the new ones with the new name `cluster2` is created, however, `cluster2` is now throwing this error `DNSResolver: Cannot resolve host: cluster1-0-0-0-...`.

As I said, there is a replicated merge tree on the cluster with the table definition something like this:

```
CREATE TABLE IF NOT EXISTS cluster1.test_table ON CLUSTER 'cluster1' (
  inserted                DateTime('UTC') COMMENT 'Inserted timestamp' CODEC(DoubleDelta, LZ4)
  )
  Engine=ReplicatedSummingMergeTree('/clickhouse/tables/cluster1/{shard}/test_table', '{replica}')
  PARTITION BY (toStartOfHour(inserted))
  TTL inserted + INTERVAL 1 DAY;
```

To ensure a smooth transition, I am confirming that no data is being replicated before changing the layout, effectively ensuring an input cardinality of 0 during the layout modification.

If i investigate the system tables such as `system.zookeeper`, `system.distributed_ddl_queue` as well as zookeeper nodes, i can still see some traces of the deleted cluster `cluster1` exists in those places.

Considering the presence of a replicated merge tree and the potential fluctuation in the number of replicas based on the new layout, or completely deleting a cluster, what is the safest approach to altering the cluster layout?

What is the safest and the correct way of deleting a cluster completely?

Thanks in advances,

**Comment by tolgaio** [comment]
or is it in my best interest to drop all the tables defined on the cluster before deleting it? so that dropping the tables first will make sure there is no residue left in any of those system tables as well as the zookeeper nodes?

**Comment by tolgaio** [comment]
I have discovered a solution to my question. Essentially, I encountered an issue with the ddl tasks residing in zookeeper. If a cluster is shut down without adequately cleaning up those tasks, it will invariably lead to problems. This is because when a new cluster is deployed using the same installation, it connects to the same zookeeper and attempts to execute all the DDLs.

As a solution, I deleted all the problematic DDL tasks from the zookeeper and the problem is solved.
